### ISSUE TITLE
Collectionextract function in C api

### ISSUE THREAD

**Issue by theroggy** [libgeos/geos#1015]
Maybe I looked over it, but I don't seem to find an [ST_CollectionExtract](https://postgis.net/docs/ST_CollectionExtract.html) equivalent in the C API.

If it is indeed not available, it would be great if it could be added...

Regards,
Pieter

**Comment by mwtoews** [comment]
The C++ API has a [`geos::geom::util::GeometryExtracter` class](https://libgeos.org/doxygen/classgeos_1_1geom_1_1util_1_1GeometryExtracter.html) (example use with [GeometryExtracterTest.cpp](https://github.com/libgeos/geos/blob/0aef713ac930e7247c50a1ae720c36f0f0bf790a/tests/unit/geom/util/GeometryExtracterTest.cpp)), but this does appear to be absent from the C API.

**Comment by dr-jts** [comment]
> If it is indeed not available, it would be great if it could be added...

What is your use case for this?  Is it to deal with mixed-dimension results from overlay operations, or something else?

**Comment by dbaston** [comment]
Related: https://trac.osgeo.org/geos/ticket/1117

**Comment by dr-jts** [comment]
Here's a possible design, for comment:
```
Geometry*    GEOSGeom_extractCollection(Geometry* collection, unsigned int dimension)
```

Should this have simple copy semantics? Or should it minimize allocations via:

* adds the elements of required dimension to a new `Multi` geometry (should this be returned as an atomic geometry if only one element is found?)
* ?? any elements not returned are destroyed automatically
* ?? caller must destroy original geometry

**Comment by pramsey** [comment]
Actually the question of "return `POLYGON()` or return `MULTIPOLYGON()`" for single-item collections seems a little fraught... doing so over a set could result in a heterogeneous set of POLYGON/MULTIPOLYGON instead of a nice homogeneous set of MULTIPOLYGON.

**Comment by dr-jts** [comment]
> Actually the question of "return `POLYGON()` or return `MULTIPOLYGON()`" for single-item collections seems a little fraught... doing so over a set could result in a heterogeneous set of POLYGON/MULTIPOLYGON instead of a nice homogeneous set of MULTIPOLYGON.

Fraught indeed...  also, extracting an arbitrary set of polygons into a MULTIPOLYGON might produce an invalid geometry, if they happen to overlap.

I suggest taking the low road and just banging all polygons into a single  MULTIPOLYGON, _caveat usor_.  For more refined semantics the user will have to write their own code.

**Comment by theroggy** [comment]
> > If it is indeed not available, it would be great if it could be added...
> 
> What is your use case for this? Is it to deal with mixed-dimension results from overlay operations, or something else?

Yes indeed. Remove eg. lines and points from collections returned from overlay operations like intersection.

**Comment by theroggy** [comment]
I suppose the GeometryExtracter mentioned above already takes this in account, but to be sure: geometrycollections can be nested, so possibly the extraction needs to be recursive (or maybe/hopefully there is some limit on the number of nestings?). 

E.g. `GEOMETRYCOLLECTION(GEOMETRYCOLLECTION(LINESTRING(0 0, 1 1)),LINESTRING(2 2, 3 3))`

**Comment by dr-jts** [comment]
> geometrycollections can be nested, so possibly the extraction needs to be recursive (or maybe/hopefully there is some limit on the number of nestings?).

Yes, this function needs to extract all elements from nested GeometryCollections.
